cluster/src/main/java/org/apache/iotdb/cluster/client/ClientManager.java

```diff
@@ -0,0 +1,203 @@
+        break;
+    }
+  }
+
+  private void constructSyncClientMap(Type type) {
+    switch (type) {
+      case ClusterClient:
+        syncClientPoolMap.put(
+            ClientCategory.DATA,
+            ClientPoolFactory.getInstance().createSyncDataPool(ClientCategory.DATA));
+        break;
+      case MetaGroupClient:
+        syncClientPoolMap.put(
+            ClientCategory.META,
+            ClientPoolFactory.getInstance().createSyncMetaPool(ClientCategory.META));
+        syncClientPoolMap.put(
+            ClientCategory.META_HEARTBEAT,
+            ClientPoolFactory.getInstance().createSyncMetaPool(ClientCategory.META_HEARTBEAT));
+        break;
+      case DataGroupClient:
+        syncClientPoolMap.put(
+            ClientCategory.DATA,
+            ClientPoolFactory.getInstance().createSyncDataPool(ClientCategory.DATA));
+        syncClientPoolMap.put(
+            ClientCategory.DATA_HEARTBEAT,
+            ClientPoolFactory.getInstance().createSyncDataPool(ClientCategory.DATA_HEARTBEAT));
+        break;
+      default:
+        logger.warn("unsupported ClientManager type: {}", type);
+        break;
+    }
+  }
+
+  /**
+   * It's safe to convert: 1. RaftService.AsyncClient to TSDataService.AsyncClient when category is
+   * DATA or DATA_HEARTBEAT; 2. RaftService.AsyncClient to TSMetaService.AsyncClient when category
+   * is META or META_HEARTBEAT.
+   *
+   * @param category
+   * @return RaftService.AsyncClient
+   */
+  @Override
+  public RaftService.AsyncClient borrowAsyncClient(Node node, ClientCategory category) {
+    try {
+      RaftService.AsyncClient client = asyncClientPoolMap.get(category).borrowObject(node);
+      if (ClientCategory.DATA == category
+          || ClientCategory.DATA_HEARTBEAT == category
+          || ClientCategory.SINGLE_MASTER == category) {
+        ((AsyncDataClient) client).setClientPool(this);
+      } else {
+        ((AsyncMetaClient) client).setClientPool(this);
+      }
+      return client;
+    } catch (NullPointerException e) {
+      logger.error("No AsyncClient pool found for {}", category, e);
+    } catch (TException e) {
+      logger.error("AsyncClient transport error for {}", category, e);
+    } catch (Exception e) {
+      logger.error("AsyncClient error for {}", category, e);
+    }
+    return null;
+  }
+
+  /**
+   * It's safe to convert: 1. RaftService.Client to TSDataService.Client when category is DATA or
+   * DATA_HEARTBEAT; 2. RaftService.Client to TSMetaService.Client when category is META or
+   * META_HEARTBEAT.
+   *
+   * @param category
+   * @return RaftService.Client
+   */
+  @Override
+  public RaftService.Client borrowSyncClient(Node node, ClientCategory category) {
+    try {
+      RaftService.Client client = syncClientPoolMap.get(category).borrowObject(node);
+      if (ClientCategory.DATA == category || ClientCategory.DATA_HEARTBEAT == category) {
+        ((SyncDataClient) client).setClientPool(this);
```

Comment:
Why are we setting clientPool all the times? Can we just set it once when creating?

Reply:
It's good idea to move the setting into construction of client. The original design is remove pool from client, that's client doesn't need to know about pool. By the design, there are to many changes in the refactor. So I add the pool manager back by `ClientManager`.